lib/ngx/pipe.md

```diff
@@ -0,0 +1,531 @@
+The timeout threshold of this reading operation can be controlled by
+[set_timeouts](#set_timeouts). The default timeout is 10 seconds.
+
+In case of success, it returns the data received; otherwise it returns three values: `nil`,
+a string describing the error, and the partial data received so far.
+
+Only one light thread is allowed to read from a process's stderr or stdout stream at a time.
+If another thread tries to read from the same stream, it will return `nil` and the error string
+`"pipe busy read"`.
+
+Each stream for stdout and stderr are separate, so you can have at most two
+light threads reading from a sub-process, one from the stdout stream and the other
+from stderr.
+
+Please note that when the `merge_stderr` option is specified in the
+[spawn](#spawn) method call,
+only one light thread can read from the sub-process, as stderr is now identical to stdout
+in this case.
+
+You can read from a data stream of the sub-process in a light thread when there is
+another light thread is pending writing to the same stream of the same
+sub-process. One light thread can, however, read from a stream while
+another light thread is *writing* to the same stream since every stream supports
+full-duplexing.
+
+Reading from an exited process's streams will return `nil` plus `closed`.
+
+[Back to TOC](#table-of-contents)
+
+stdout_read_line
+----------------
+**syntax:** *data, err, partial = proc:stdout_read_line()*
+
+**context:** *phases support yielding*
+
+Similar to [stderr_read_line](#stderr_read_line) but working on the sub-process's
+stdout stream.
+
+[Back to TOC](#table-of-contents)
+
+stderr_read_line
+----------------
+**syntax:** *data, err, partial = proc:stderr_read_line()*
+
+**context:** *phases support yielding*
+
+Reads data like [stderr_read_all](#stderr_read_all) but only reads a single line of data.
+
+When the data stream is truncated without a new-line character, it returns 3 values:
+`nil`, the error string `"closed"`, and the partial data received so far.
+
+The line should be terminated by a `Line Feed` (LF) character (ASCII 10),
+optionally preceded by a `Carriage Return` (CR) character (ASCII 13).
+The CR and LF characters are not included in the returned line data.
+
+[Back to TOC](#table-of-contents)
+
+stdout_read_bytes
+-----------------
+
+**syntax:** *data, err, partial = proc:stdout_read_bytes(len)*
+
+**context:** *phases support yielding*
+
+Similar to [stderr_read_bytes](#stderr_read_bytes) but working on the sub-process's
+stdout stream.
+
+[Back to TOC](#table-of-contents)
+
+stderr_read_bytes
+-----------------
+**syntax:** *data, err, partial = proc:stderr_read_bytes(len)*
+
+**context:** *phases support yielding*
+
+Reads data from stderr like [stderr_read_all](#stderr_read_all) but only reads the
+specified number of bytes of data.
+
+If the data stream is truncated with less bytes of data available, it returns
+3 values: `nil`, the error string `"closed"`, and the partial data string received so far.
+
```

Comment:
Should we put the docs of the `stderr_read_xxx` before `stdout_read_xxx`? We should put the real docs first, I think?